text/0126-introduce-xcq.md

```diff
@@ -0,0 +1,342 @@
+- Gas computation: According to [this issue](https://github.com/koute/polkavm/issues/17), the gas cost model of PolkaVM is not accurate for now.
+
+## Testing, Security, and Privacy
+
+- Testing:
+  - A comprehensive test suite should be developed to cover various scenarios:
+    - Positive test cases:
+      - Basic queries with various extensions, data types, return values, custom computations, etc.
+      - Accurate conversion between given weight limit and the gas limit of PolkaVM
+    - Negative test cases:
+      - Queries exceeding weight limits
+      - Invoking queries from unauthorized sources
+    - Edge cases:
+      - Queries with minimal or maximum allowed input sizes
+  - Integration tests to ensure proper interaction with off-chain wallets/UI and on-chain XCM, including the aforementioned use cases in **Motivation** section.
+
+- Security:
+  - The XCQ system must enforce a strict read-only policy for all query operations. A mechanism should be implemented to prevent any state-changing operations within XCQ queries.
+  - Clear guidelines and best practices should be provided for parachain developers to ensure secure implementation.
+
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+It's a new functionality, which doesn't modify the existing implementations.
+
+### Ergonomics
+
+The proposal facilitate the wallets and dApps developers. Developers no longer need to examine every concrete implementation to support conceptually similar operations across different chains. Additionally, they gain a more modular development experience through encapsulating custom computations over the exposed APIs in PolkaVM programs.
+### Compatibility
+
+The proposal defines new apis, which doesn't break compatibility with existing interfaces.
+
+## Prior Art and References
+
+There are several discussions related to the proposal, including:
+
+- <https://forum.polkadot.network/t/wasm-view-functions/1045> is the original discussion about having a mechanism to avoid code duplications between the runtime and front-ends/wallets. In the original design, the custom computations are compiled as a wasm function.
+- <https://github.com/paritytech/polkadot-sdk/issues/216> is the issue tracking the `view functions` implementation in runtime implementations
+- <https://github.com/paritytech/polkadot-sdk/pull/4722> is the on-going `view function` pull request. It works at pallet level. If two chains use two different pallets to provide similar functionalities, like pallet-assets and pallet-erc20, we still need to have different codes to support. Therefore, it doesn't conflict with XCQ, and can be utilized by XCQ.
+
+## Unresolved Questions
+
+
+## Future Directions and Related Material
+
+Since XCQ are supported both in off-chain and on-chain, a related XCM-Format RFC should be proposed.
```

Comment:
We recently merged XCM RFCs into the fellowship RFCs. So the new instruction and behaviour of XCM could be described here.